_posts/2020-04-18-regression-chains/regression-chains.Rmd

```diff
@@ -0,0 +1,179 @@
+---
+title: "Regression chains"
+categories:
+  - regression
+  - mlr3pipelines
+author:
+  - name: Lennart Schneider
+date: 04-18-2020
+description: |
+  We show how to use mlr3pipelines to do regression chains.
+output:
+  distill::distill_article:
+    self_contained: false
+    css: ../../custom.css
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(
+  echo = TRUE,
+  R.options = list(width = 80)
+)
+library(mlr3book)
+lgr::get_logger("mlr3")$set_threshold("warn")
+```
+
+In this tutorial we demonstrate how to use `r mlr_pkg("mlr3pipelines")` to handle multi-target regression by arranging regression models into a chain, i.e., creating a linear sequence of regression models.
+
+## Regression Chains
+
+In a simple regression chain, regression models are arranged in a linear sequence. Here, the first model will use the input to predict a single output and the second model will use the input and the output of the first model to make its own prediction and so on. For more details, see e.g. [Spyromitros-Xioufis, Tsoumakas, Groves & Vlahavas (2016)](https://doi.org/10.1007/s10994-016-5546-z).
+
+## Before you start
+
+The following sections describe an approach towards working with `Task`s that have multiple targets.
+E.g. in the example below, we have 3 target variables **y1** to **y3**.
+This type of Task can be created via the [mlr3multioutput](https://github.com/mlr-org/mlr3multioutput/) (currently under development) in the future.
+The current goal of this post is to show how such modeling steps can be written as a relatively small amount of pipeline steps and how such steps can be put together. Writing pipelines with such steps allows for great flexibility in modelling more complicated scenarios such as the ones described below.
+
+## Data
+
+In the following, we rely on some toy data. We simulate 100 responses to three target variables, $y_{1}$, $y_{2}$, and $y_{3}$ following a multivariate normal distribution with a mean and covariance matrix of:
+
+```{r}
+library(data.table)
+library(mvtnorm)
+set.seed(2409)
+n = 100
+(mean = c(y1 = 1, y2 = 2, y3 = 3))
+(sigma = matrix(c(1, -0.5, 0.25, -0.5, 1, -0.25, 0.25, -0.25, 1),
+  nrow = 3, ncol = 3, byrow = TRUE))
+Y = rmvnorm(n, mean = mean, sigma = sigma)
+```
+
+The feature variables $x_{1}$, and $x_{2}$ are simulated as follows: $x_{1}$ is simply given by $y_{1}$ and an independent standard normally distributed error term and $x_{2}$ is given by $y_{2}$ and an independent standard normally distributed error term.
+
+```{r}
+x1 = Y[, 1] + rnorm(n)
+x2 = Y[, 2] + rnorm(n)
+```
+
+The final data is given as:
+
+```{r}
+dat = as.data.table(cbind(Y, x1, x2))
+str(dat)
+```
+
```

Comment:
This simulates a situation, where we have multiple target variables that are correlated with each other, such that predicting them along with each other can improve the resulting prediction model.

A common example for such a situation is e.g. hospital data, where time spent in the ICU (not known a priori), or whether a patient will likely die heavily influence the cost incurred by a patient's treatment.